File: backend/internal/server/dto/riskcontrol.go
package dto

import (
	appriskcontrol "github.com/DouDOU-start/airgate-core/internal/app/riskcontrol"
	"github.com/DouDOU-start/airgate-core/internal/moderation"
)

// RiskControlConfigResp 风控配置回显（审核 key 只出掩码与健康状态）。
type RiskControlConfigResp struct {
	RiskControlEnabled   bool                   `json:"risk_control_enabled"`
	Enabled              bool                   `json:"enabled"`
	Mode                 string                 `json:"mode"`
	BaseURL              string                 `json:"base_url"`
	Model                string                 `json:"model"`
	APIKeyCount          int                    `json:"api_key_count"`
	APIKeyMasks          []string               `json:"api_key_masks"`
	APIKeyStatuses       []moderation.KeyStatus `json:"api_key_statuses"`
	TimeoutMS            int                    `json:"timeout_ms"`
	SampleRate           int                    `json:"sample_rate"`
	AllGroups            bool                   `json:"all_groups"`
	GroupIDs             []int                  `json:"group_ids"`
	RecordNonHits        bool                   `json:"record_non_hits"`
	Thresholds           map[string]float64     `json:"thresholds"`
	Categories           []string               `json:"categories"`
	WorkerCount          int                    `json:"worker_count"`
	QueueSize            int                    `json:"queue_size"`
	BlockStatus          int                    `json:"block_status"`
	BlockMessage         string                 `json:"block_message"`
	EmailOnHit           bool                   `json:"email_on_hit"`
	AutoBanEnabled       bool                   `json:"auto_ban_enabled"`
	BanThreshold         int                    `json:"ban_threshold"`
	ViolationWindowHours int                    `json:"violation_window_hours"`
	RetryCount           int                    `json:"retry_count"`
	HitRetentionDays     int                    `json:"hit_retention_days"`
	NonHitRetentionDays  int                    `json:"non_hit_retention_days"`
	PreHashCheckEnabled  bool                   `json:"pre_hash_check_enabled"`
	BlockedKeywords      []string               `json:"blocked_keywords"`
	KeywordBlockingMode  string                 `json:"keyword_blocking_mode"`
	ModelFilter          moderation.ModelFilter `json:"model_filter"`
}

// UpdateRiskControlConfigReq 配置增量更新请求（缺省字段不改动）。
type UpdateRiskControlConfigReq struct {
	RiskControlEnabled   *bool                   `json:"risk_control_enabled"`
	Enabled              *bool                   `json:"enabled"`
	Mode                 *string                 `json:"mode"`
	BaseURL              *string                 `json:"base_url"`
	Model                *string                 `json:"model"`
	APIKeys              *[]string               `json:"api_keys"`
	APIKeysMode          string                  `json:"api_keys_mode"`
	DeleteAPIKeyHashes   []string                `json:"delete_api_key_hashes"`
	ClearAPIKeys         bool                    `json:"clear_api_keys"`
	TimeoutMS            *int                    `json:"timeout_ms"`
	SampleRate           *int                    `json:"sample_rate"`
	AllGroups            *bool                   `json:"all_groups"`
	GroupIDs             *[]int                  `json:"group_ids"`
	RecordNonHits        *bool                   `json:"record_non_hits"`
	Thresholds           *map[string]float64     `json:"thresholds"`
	WorkerCount          *int                    `json:"worker_count"`
	QueueSize            *int                    `json:"queue_size"`
	BlockStatus          *int                    `json:"block_status"`
	BlockMessage         *string                 `json:"block_message"`
	EmailOnHit           *bool                   `json:"email_on_hit"`
	AutoBanEnabled       *bool                   `json:"auto_ban_enabled"`
	BanThreshold         *int                    `json:"ban_threshold"`
	ViolationWindowHours *int                    `json:"violation_window_hours"`
	RetryCount           *int                    `json:"retry_count"`
	HitRetentionDays     *int                    `json:"hit_retention_days"`
	NonHitRetentionDays  *int                    `json:"non_hit_retention_days"`
	PreHashCheckEnabled  *bool                   `json:"pre_hash_check_enabled"`
	BlockedKeywords      *[]string               `json:"blocked_keywords"`
	KeywordBlockingMode  *string                 `json:"keyword_blocking_mode"`
	ModelFilter          *moderation.ModelFilter `json:"model_filter"`
}

// TestRiskControlKeysReq 审核 key 探活/试审请求。
type TestRiskControlKeysReq struct {
	APIKeys   []string `json:"api_keys"`
	BaseURL   string   `json:"base_url"`
	Model     string   `json:"model"`
	TimeoutMS int      `json:"timeout_ms"`
	Prompt    string   `json:"prompt"`
	Images    []string `json:"images"`
}

// RiskControlLogsQuery 审核日志查询参数。
type RiskControlLogsQuery struct {
	PageReq
	Result   string `form:"result" binding:"omitempty,oneof=hit blocked pass error"`
	GroupID  string `form:"group_id"`
	Endpoint string `form:"endpoint"`
	Search   string `form:"search"`
	From     string `form:"from"`
	To       string `form:"to"`
}

// RiskControlLogResp 审核日志行（app 层 Record 自带序列化标签，直接复用）。
type RiskControlLogResp = appriskcontrol.Record

// RiskControlUnbanResp 解封结果。
type RiskControlUnbanResp struct {
	UserID int    `json:"user_id"`
	Status string `json:"status"`
}

// DeleteRiskControlHashReq 删除单条命中哈希请求。
type DeleteRiskControlHashReq struct {
	InputHash string `json:"input_hash" binding:"required"`
}

// DeleteRiskControlHashResp 删除结果。
type DeleteRiskControlHashResp struct {
	InputHash string `json:"input_hash"`
	Deleted   bool   `json:"deleted"`
}

// ClearRiskControlHashesResp 清空结果。
type ClearRiskControlHashesResp struct {
	Deleted int64 `json:"deleted"`
}

File: web/src/shared/api/riskControl.ts
import { get, post, put, del } from './client';
import type { PagedData } from '../types';

// ==================== 类型 ====================

export type ModerationMode = 'off' | 'observe' | 'pre_block';
export type KeywordBlockingMode = 'keyword_only' | 'keyword_and_api' | 'api_only';
export type ModelFilterType = 'all' | 'include' | 'exclude';
export type ModerationKeyStatusValue = 'unknown' | 'ok' | 'error' | 'frozen';

export interface ModerationModelFilter {
  type: ModelFilterType;
  models: string[];
}

export interface ModerationKeyStatus {
  index: number;
  key_hash: string;
  masked: string;
  status: ModerationKeyStatusValue;
  failure_count: number;
  success_count: number;
  last_error: string;
  last_checked_at?: string;
  frozen_until?: string;
  last_latency_ms: number;
  last_http_status: number;
  last_tested: boolean;
  configured: boolean;
}

export interface ModerationKeyLoad {
  index: number;
  key_hash: string;
  masked: string;
  status: ModerationKeyStatusValue;
  active: number;
  total: number;
  success: number;
  errors: number;
  avg_latency_ms: number;
  last_latency_ms: number;
  last_http_status: number;
}

export interface RiskControlConfig {
  mode: ModerationMode;
  base_url: string;
  model: string;
  api_key_count: number;
  api_key_masks: string[];
  api_key_statuses: ModerationKeyStatus[];
  timeout_ms: number;
  sample_rate: number;
  all_groups: boolean;
  group_ids: number[];
  record_non_hits: boolean;
  thresholds: Record<string, number>;
  categories: string[];
  worker_count: number;
  queue_size: number;
  block_status: number;
  block_message: string;
  email_on_hit: boolean;
  auto_ban_enabled: boolean;
  ban_threshold: number;
  violation_window_hours: number;
  retry_count: number;
  hit_retention_days: number;
  non_hit_retention_days: number;
  pre_hash_check_enabled: boolean;
  blocked_keywords: string[];
  keyword_blocking_mode: KeywordBlockingMode;
  model_filter: ModerationModelFilter;
}

export interface UpdateRiskControlConfigReq {
  mode?: ModerationMode;
  base_url?: string;
  model?: string;
  api_keys?: string[];
  api_keys_mode?: 'append' | 'replace';
  delete_api_key_hashes?: string[];
  clear_api_keys?: boolean;
  timeout_ms?: number;
  sample_rate?: number;
  all_groups?: boolean;
  group_ids?: number[];
  record_non_hits?: boolean;
  thresholds?: Record<string, number>;
  worker_count?: number;
  queue_size?: number;
  block_status?: number;
  block_message?: string;
  email_on_hit?: boolean;
  auto_ban_enabled?: boolean;
  ban_threshold?: number;
  violation_window_hours?: number;
  retry_count?: number;
  hit_retention_days?: number;
  non_hit_retention_days?: number;
  pre_hash_check_enabled?: boolean;
  blocked_keywords?: string[];
  keyword_blocking_mode?: KeywordBlockingMode;
  model_filter?: ModerationModelFilter;
}

export interface RiskControlStatus {
  mode: ModerationMode | '';
  worker_count: number;
  max_workers: number;
  active_workers: number;
  queue_size: number;
  queue_length: number;
  queue_usage_percent: number;
  enqueued: number;
  dropped: number;
  processed: number;
  errors: number;
  pre_block_active: number;
  pre_block_checked: number;
  pre_block_allowed: number;
  pre_block_blocked: number;
  pre_block_errors: number;
  pre_block_avg_latency_ms: number;
  api_key_available_count: number;
  api_key_loads: ModerationKeyLoad[];
  api_key_statuses: ModerationKeyStatus[];
  flagged_hash_count: number;
  last_cleanup_at?: string;
  last_cleanup_deleted_hit: number;
  last_cleanup_deleted_non_hit: number;
}

export interface TestRiskControlKeysReq {
  api_keys?: string[];
  base_url?: string;
  model?: string;
  timeout_ms?: number;
  prompt?: string;
  images?: string[];
}

export interface RiskControlTestAuditResult {
  flagged: boolean;
  highest_category: string;
  highest_score: number;
  category_scores: Record<string, number>;
  thresholds: Record<string, number>;
}

export interface TestRiskControlKeysResp {
  items: ModerationKeyStatus[];
  audit_result?: RiskControlTestAuditResult;
  image_count: number;
}

export interface ModerationLog {
  id: number;
  request_id: string;
  user_id: number;
  user_email: string;
  api_key_id: number;
  group_id: number;
  group_name: string;
  endpoint: string;
  protocol: string;
  model: string;
  mode: string;
  action: string;
  flagged: boolean;
  highest_category: string;
  highest_score: number;
  matched_keyword: string;
  category_scores: Record<string, number>;
  threshold_snapshot: Record<string, number>;
  input_excerpt: string;
  input_hash: string;
  upstream_latency_ms: number;
  queue_delay_ms: number;
  error: string;
  violation_count: number;
  auto_banned: boolean;
  email_sent: boolean;
  user_status: string;
  created_at: string;
}

export interface RiskControlLogsQuery {
  page?: number;
  page_size?: number;
  result?: 'hit' | 'blocked' | 'pass' | 'error' | '';
  group_id?: number;
  endpoint?: string;
  search?: string;
  from?: string;
  to?: string;
}

// ==================== API ====================

export const riskControlApi = {
  getConfig: () => get<RiskControlConfig>('/api/v1/admin/risk-control/config'),
  updateConfig: (data: UpdateRiskControlConfigReq) =>
    put<RiskControlConfig>('/api/v1/admin/risk-control/config', data),
  getStatus: () => get<RiskControlStatus>('/api/v1/admin/risk-control/status'),
  testKeys: (data: TestRiskControlKeysReq) =>
    post<TestRiskControlKeysResp>('/api/v1/admin/risk-control/api-keys/test', data),
  listLogs: (params?: RiskControlLogsQuery) =>
    get<PagedData<ModerationLog>>('/api/v1/admin/risk-control/logs', params),
  unbanUser: (userId: number) =>
    post<{ user_id: number; status: string }>(`/api/v1/admin/risk-control/users/${userId}/unban`, {}),
  deleteHash: (inputHash: string) =>
    del<{ input_hash: string; deleted: boolean }>('/api/v1/admin/risk-control/hashes', { input_hash: inputHash }),
  clearLogs: (result?: string) =>
    del<{ deleted: number; result: string }>('/api/v1/admin/risk-control/logs', { result: result ?? '' }),
  clearHashes: () =>
    del<{ deleted: number }>('/api/v1/admin/risk-control/hashes/all'),
};
